File: src/main/java/org/dxworks/utils/java/rest/client/utils/JsonMapper.java
package org.dxworks.utils.java.rest.client.utils;

import com.google.api.client.json.JsonGenerator;
import com.google.api.client.json.JsonObjectParser;
import com.google.api.client.json.jackson2.JacksonFactory;
import lombok.SneakyThrows;

import java.io.*;
import java.lang.reflect.Type;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

public class JsonMapper {

    private final JacksonFactory jsonFactory = new JacksonFactory();

    public <T> T readJSON(String jsonString, Class<T> type) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(new StringReader(jsonString), type);
    }

    public <T> T readJSONfromFile(File file, Class<T> type) throws IOException {
        return readJSONfromFile(file, type, Charset.defaultCharset());
    }

    public <T> T readJSONfromFile(File file, Class<T> type, Charset encoding) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(getReader(file, encoding), type);
    }

    public <T> T readJSON(Reader reader, Class<T> type) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(reader, type);
    }

    public Object readJSON(String jsonString, Type type) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(new StringReader(jsonString), type);
    }

    public Object readJSONfromFile(File file, Type type) throws IOException {
        return readJSONfromFile(file, type, Charset.defaultCharset());
    }

    public Object readJSONfromFile(File file, Type type, Charset encoding) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(getReader(file, encoding), type);
    }

    private InputStreamReader getReader(File file, Charset encoding) throws FileNotFoundException {
        return new InputStreamReader(new FileInputStream(file), encoding);
    }

    public Object readJSON(Reader reader, Type type) throws IOException {
        return new JsonObjectParser(jsonFactory).parseAndClose(reader, type);
    }

    public String getSerializedObject(Object o) throws IOException {
        Writer out = new StringWriter();
        writeJSON(out, o);
        return out.toString();
    }

    public String getPrettySerializedObject(Object o) throws IOException {
        Writer out = new StringWriter();
        writePrettyJSON(out, o);
        return out.toString();
    }

    public void writePrettyJSON(Writer writer, Object o) throws IOException {
        JsonGenerator generator = jsonFactory.createJsonGenerator(writer);
        generator.enablePrettyPrint();
        generator.serialize(o);
        generator.flush();
    }

    public void writeJSON(Writer writer, Object o) throws IOException {
        JsonGenerator generator = jsonFactory.createJsonGenerator(writer);
        generator.serialize(o);
        generator.flush();
    }

    public void writePrettyJSONtoFile(File file, Object o) throws IOException {
        writePrettyJSONtoFile(file, o, Charset.defaultCharset());
    }

    public void writePrettyJSONtoFile(File file, Object o, Charset encoding) throws IOException {
        JsonGenerator generator = jsonFactory.createJsonGenerator(getWriter(file));
        generator.enablePrettyPrint();
        generator.serialize(o);
        generator.flush();
    }

    public void writeJSONtoFile(File file, Object o) throws IOException {
        writeJSONtoFile(file, o, Charset.defaultCharset());
    }

    public void writeJSONtoFile(File file, Object o, Charset encoding) throws IOException {
        JsonGenerator generator = jsonFactory.createJsonGenerator(getWriter(file));
        generator.serialize(o);
        generator.flush();
    }

    @SneakyThrows
    private Writer getWriter(File file) {
        return new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8);
    }
}

File: src/main/kotlin/org/dxworks/utils/java/rest/client/utils/JsonMapper.kt
package org.dxworks.utils.java.rest.client.utils

import com.google.api.client.json.JsonObjectParser
import com.google.api.client.json.jackson2.JacksonFactory
import java.io.*
import java.lang.reflect.Type
import java.nio.charset.Charset

open class JsonMapper {
    private val jsonFactory = JacksonFactory()

    fun <T> readJSON(jsonString: String, type: Class<T>): T {
        return JsonObjectParser(jsonFactory).parseAndClose(StringReader(jsonString), type)
    }

    fun <T> readJSON(reader: Reader, type: Class<T>): T {
        return JsonObjectParser(jsonFactory).parseAndClose(reader, type)
    }

    fun <T> readJSONfromFile(file: File, type: Class<T>, encoding: Charset = Charset.defaultCharset()): T {
        return JsonObjectParser(jsonFactory).parseAndClose(getReader(file, encoding), type)
    }

    fun readJSON(jsonString: String, type: Type): Any {
        return JsonObjectParser(jsonFactory).parseAndClose(StringReader(jsonString), type)
    }

    fun readJSON(reader: Reader, type: Type): Any {
        return JsonObjectParser(jsonFactory).parseAndClose(reader, type)
    }

    fun readJSONfromFile(file: File, type: Type, encoding: Charset = Charset.defaultCharset()): Any {
        return JsonObjectParser(jsonFactory).parseAndClose(getReader(file, encoding), type)
    }

    private fun getReader(file: File, encoding: Charset): InputStreamReader {
        return InputStreamReader(FileInputStream(file), encoding)
    }

    fun getSerializedObject(o: Any): String {
        return StringWriter().also {
            writeJSON(it, o)
        }.toString()
    }

    fun getPrettySerializedObject(o: Any): String {
        return StringWriter().also {
            writePrettyJSON(it, o)
        }.toString()
    }

    fun writePrettyJSON(writer: Writer, o: Any) {
        val generator = jsonFactory.createJsonGenerator(writer).apply {
            enablePrettyPrint()
            serialize(o)
            flush()
        }
    }

    fun writeJSON(writer: Writer, o: Any) {
        jsonFactory.createJsonGenerator(writer).apply {
            serialize(o)
            flush()
        }
    }

    fun writePrettyJSONtoFile(file: File, o: Any, encoding: Charset = Charset.defaultCharset()) {
        jsonFactory.createJsonGenerator(getWriter(file, encoding)).apply {
            enablePrettyPrint()
            serialize(o)
            flush()
        }
    }

    fun writeJSONtoFile(file: File, o: Any, encoding: Charset = Charset.defaultCharset()) {
        jsonFactory.createJsonGenerator(getWriter(file, encoding)).apply {
            serialize(o)
            flush()
        }
    }

    private fun getWriter(file: File, encoding: Charset): Writer {
        return OutputStreamWriter(FileOutputStream(file), encoding)
    }
}
